### DISCUSSION TITLE
Access filter treats undefined as allow all

### DISCUSSION THREAD

**Discussion by christopherliu** [keystonejs/keystone#8383] [Questions]
I'm using the access filter function ([documented here](https://keystonejs.com/docs/guides/auth-and-access-control#how-filter-access-control-works)) and have found that passing `undefined` results in no restrictions on retrieved items.

Snippet of code below:

```typescript
export const UserProgram = list({
    access: {
        filter: {
            query: isAdminOrOwner,
//...


function isAdminOrOwner({ session, context, listKey, operation }) {
    return isAdmin({ session }) || {
        student: {
            id: {
                equals: session?.data.id
            }
        }
    };
}
```

When a user is not logged in, `session` is empty, and `session?.data.id` is `undefined`. According to the documentation:

> An easy way to set up Access Control filters is to write them as queries in the GraphQL Playground, then test them to make sure they return the results you expect. When you're happy, copy and paste the where input into your access control function.

However, when the filter is set to `equals: undefined`, instead of failing as the code in the GraphQL Playground does, the access filter silently allows all rows to be retrieved.

Using:

- Keystone 6, `"@keystone-6/core": "^2.1.0",` in package.json.

Thanks!

**Comment by borisno2** [comment]
Hi @christopherliu

The filter acccess control functions should return a graphql query or a boolean, if they return a GraphQL query this is used to filter the results in Keystone's Graphql Resolver. When you have `{ id: { equals: undefined } }` this is turns into `{ id: { } }` as `equals` is `undefined`.  This is still a valid GraphQL where clause, but will will not filter out any items. In my testing in the playground, a validation error is given (`value expected` in the playground), but the code still runs and returns all the items in the list.

I suggest you use `access.operation` in combination with `access.filter` (`access.operation` is required in the latest versions of `@keystone-6/core` to help with these situations) to check if there is a valid session and return `false` if there is no `session`. 

The [auth example](https://github.com/keystonejs/keystone/blob/main/examples/auth) has some exmaples of using `access.filter` 
Thanks

**Reply by christopherliu** [reply]
Thanks! I was able to do something similar. It wouldn't prevent me from passing an undefined if I omit the `access.operation` by accident, though.

**Comment by dcousens** [comment]
@christopherliu are you using typescript?
If your `session` isn't `any`, you could use the Typescript configuration option `"exactOptionalPropertyTypes": true` to prevent `undefined` from being allowed.

**Reply by christopherliu** [reply]
Thanks! Where do I specify that? Does Keystone.js accept a `tsconfig.json`? Adding a file named tsconfig.json with that option turned on didn't seem to block me using `undefined` the way it would if I were using `tsc`.

Separately, while VS Code shows `filter.query` to be type `ListFilterAccessControl<"query", BaseListTypeInfo>?`, passing invalid values to `student.id.equals` doesn't seem to bother the type checker, while passing invalid values to `filter.query` directly does.

**Reply by dcousens** [reply]
@christopherliu you need to ensure your `session` isn't `any`, and your `context` type is refined.
You can add a `Context` type to your function like below

```typescript
import type { Context } from '.keystone/types'

type Session = {
  itemId: string
  data: {
    // you need to manually define this
  }
}

function isAdminOrOwner({ session, context }: {
  session: Session,
  context: Context,
}) {
```

I am hoping we can find ways to improve this DX.

**Reply by christopherliu** [reply]
Thanks @dcousens . Even with specifying a Session and a Context type, it doesn't look like the type checker complains about the filter being null. On the other hand it does complain about a mismatch between the context in types and the context expected by filter:

![image](https://user-images.githubusercontent.com/519376/225696960-5cd1bdf6-6ba4-4386-8254-2a6f2bcbf08d.png)
